### PR TITLE
WebSocket: reply with Close frame for peer-caused protocol errors

### PR DESCRIPTION
**Pull request by smerritt**

### CODE REVIEW

**Review by kentonv — COMMENTED** [review]
Nice!

I think my comments below all turn out to be not issues in practice but some minor tweaks could assuage the fears.

File: c++/src/kj/compat/http.c++ [hunk 1]
```diff
@@ -3437,12 +3437,8 @@ private:
 
     KJ_IF_SOME(p, sendingPong) {
       // We recently sent a pong, make sure it's finished before proceeding.
-      auto promise = p.then([this, opcode, message]() {
-        currentlySending = false;
-        return sendImpl(opcode, message);
-      });
+      co_await p;
```
[the remark this answers is not in the document]
Comment (smerritt) [reply c++/src/kj/compat/http.c++]: Makes sense; done. It's now a loop that checks `disconnected` on each iteration, so I think the old behavior is matched.

File: c++/src/kj/compat/http.c++ [hunk 2]
```diff
@@ -3491,21 +3487,16 @@ private:
...
-        "support an extension that would set these bits");
+	      "support an extension that would set these bits");
 
-    auto promise = stream->write(sendParts).attach(kj::mv(compressedMessage));
-    if (!mask.isZero()) {
-      promise = promise.attach(kj::mv(ownMessage));
-    }
-    return promise.then([this, size = sendParts[0].size() + sendParts[1].size()]() {
-      currentlySending = false;
+    co_await stream->write(sendParts);
```
[the remark this answers is not in the document]
Comment (smerritt) [reply c++/src/kj/compat/http.c++:3507]: Nice! Done.

File: c++/src/kj/compat/http.c++ [hunk 3]
```diff
@@ -3492,53 +3505,54 @@ private:
...
   void queuePong(kj::Array<byte> payload) {
-    bool alreadyWaitingForPongWrite = (queuedPong != kj::none);
+    bool alreadyWaitingForPongWrite = (queuedControlMessage != kj::none);
 
     // Note: According to spec, if the server receives a second ping before responding to the
     //   previous one, it can opt to respond only to the last ping. So we don't have to check if
-    //   queuedPong is already non-null.
-    queuedPong = kj::mv(payload);
+    //   queuedControlMessage is already non-null.
+    queuedControlMessage = ControlMessage(OPCODE_PONG, kj::mv(payload));
```
[the remark this answers is not in the document]
Comment (smerritt) [reply c++/src/kj/compat/http.c++]: I added some extra comments. Let me know if they're clear.

**Comment by smerritt** [comment]
Ugh, my tooling rebased onto the wrong branch. Let me un-bork that...

**Comment by smerritt** [comment]
Alright, sorry for the noise. I've got it sorted out now.

Fixups:

* 90d34df3
* f9b057a7
* 547cfebd
* bb4c310e
* db9d1b87
* fdc96958

**Comment by smerritt** [comment]
One more fixup (just extraneous blank comment lines): 9cbe9818

File: c++/src/kj/compat/http.c++ [hunk 4]
```diff
@@ -3435,14 +3441,16 @@ private:
...
-    KJ_IF_SOME(p, sendingPong) {
-      // We recently sent a pong, make sure it's finished before proceeding.
-      auto promise = p.then([this, opcode, message]() {
-        currentlySending = false;
-        return sendImpl(opcode, message);
-      });
-      sendingPong = kj::none;
-      return promise;
+    while (sendingControlMessage != kj::none) {
+      KJ_IF_SOME(p, sendingControlMessage) {
```
Comment (kentonv) [inline c++/src/kj/compat/http.c++:3445]: Nitpick, feel free to ignore:

There's some part of my brain that really doesn't like that this is checking the same thing the previous line was checking.

I think I would normally write this as `for (;;) { KJ_IF_SOME(...) { ... } else { break; } }`. Which is not pretty either but avoids the duplicate check.

That said I wonder if we should create a `KJ_WHILE_SOME`. The pattern has come up a few times.

File: c++/src/kj/compat/http.c++ [hunk 5]
```diff
@@ -3488,81 +3496,132 @@ private:
...
-    bool alreadyWaitingForPongWrite = (queuedPong != kj::none);
+    bool alreadyWaitingForPongWrite = false;
+
+    KJ_IF_SOME(controlMessage, queuedControlMessage) {
+      if (controlMessage.opcode == OPCODE_CLOSE) {
+	// We're currently sending a Close message, which we only do (at least via queuedControlMessage) when we're
+	// closing the connection due to error. There's no point queueing a Pong that'll never be sent.
+	return;
+      } else {
+	alreadyWaitingForPongWrite = true;
```
Comment (kentonv) [inline c++/src/kj/compat/http.c++:3562]: Should we `KJ_ASSERT(controlMessage.opcode == OPCODE_PONG)`?

**Review by kentonv — APPROVED** [review]

File: c++/src/kj/compat/http.c++ (hunk 4 above)
Comment (smerritt) [reply c++/src/kj/compat/http.c++:3445]: Done.

I'm in favor of `KJ_WHILE_SOME`. We've got a branching construct for dealing with `Maybe`; let's also have a loop.

File: c++/src/kj/compat/http.c++ (hunk 5 above)
Comment (smerritt) [reply c++/src/kj/compat/http.c++:3565]: Yeah, that sounds like a thing that future-us will appreciate. Done.

**Pull request merged**